### ISSUE TITLE
NamedControl doesn't see Controls created by other APIs

### ISSUE THREAD

**Issue by avdrd** [supercollider/supercollider#5648]
I'm not entirely sure if this qualifies as a bug report rather than a feature request, but since the [help page for Controls](https://doc.sccode.org/Classes/Control.html) is rather ambiguous on this, because it freely mixes old and new interfaces *creating an expectation of interoperability*...

As was discussed on the forum, NamedControl doesn't see controls created with function `arg`s nor is it interoperable with the (older?) Control.names(..).kr(...) way of dynamically instantiating controls. 

## Environment

* SuperCollider version: 3.12.1
* Operating system: all

## Steps to reproduce

As examples, first some artificial ones:

### NamedControl doesn't see arg controls

```
d = SynthDef(\one, { arg xx; Out.kr(0, [xx, \xx.kr]) }).dumpUGens
/*
[ 0_Control, control, nil ]
[ 1_Control, control, nil ]
[ 2_Out, control, [ 0, 0_Control[0], 1_Control[0] ] ]
*/
```
Two controls are created instead of `NamedControl` looking up `xx` that was created by args. Due to the way the server resolves such conflict, meaning stopping the search at the 1st maching name, the 2nd control  is not addressable by name. `SynthDesc` will actually produce a warning in such cases when it build the msgFunc, but other workflows like Ndefs don't produce any such warning but result in silent failures.

```
d.add
// -> a SynthDef
// WARNING: Could not build msgFunc for this SynthDesc: duplicate control name xx etc.

Ndef(\one, { arg xx; [xx, \xx.kr] }).prime;
// no warning whatsoever
```

### NamedControl doesn't see controls issued by Control.names.kr and friends

Mixing Control.names wtih NamedControl produces similarly problematic graphs e.g.

```
SynthDef(\one, { Out.kr(0, [Control.names([\xx]).kr([12]), \xx.kr]) }).dumpUGens
/*
[ 0_Control, control, nil ]
[ 1_Control, control, nil ]
[ 2_Out, control, [ 0, 0_Control[0], 1_Control[0] ] ]
*/
```

### JITLib uses a Control.name([\out]) internally, duplicating controls in some cases

And more realistically, [there's trouble](https://scsynth.org/t/why-is-nodeproxy-outputting-on-the-wrong-bus-if-in-kr-out-ir-is-used-inside-its-function) when NamedControl is used inside a Ndef function to try to access `out`, because JITLib uses `Contol.names` internally to instantiate another `out`.

```
// reserve these for something else; JITLib should not touch 
c = Bus.control(s, 4);

y = NodeProxy.control(s, 1);
// make some self-triggering decay as simple illustration
y[0] = { var oi = In.kr(\out.ir(42)); if(oi > 0.01, oi * 0.999, 1.0) };

y.get(\out) // 4 -- seems to respect others, but...
y.trace
```
the output from NodeProxy goes to the wrong bus. Apparently that only happens when In.kr(\out.ir) is used that way inside the proxy function. The trace prints something like
```
  unit 1 In
    in  4
    out 0
// more units omitted
 unit 10 Out
    in  0 1
    out
```
while the input is wired correctly, the output is not! Meaning NodeProxy corrupted (or overtook) a bus (number 0) reserved for something else.

## Expected vs. actual behavior
Perhaps the desirable behavior here is for a unified namespace, as much as possible.

## Fix(es) proposed

I've done some prototyping work on this.

* (Pre-)loading `arg`-generated controls in to NamedControl [doesn't take all that much](https://scsynth.org/t/preloading-synthdef-func-args-into-namedcontrols-almost-unified-control-namespace/5000?u=avid_reader) in the way of code changes.

* Updating NamedControl from dynamically generated Contrlol.names [is more complicated](https://scsynth.org/t/preloading-synthdef-func-args-into-namedcontrols-almost-unified-control-namespace/5000/2?u=avid_reader), but for the simple and common case where a single name is issued at a time with `names`, I also have some a working prototype.

On the forum, opinions have been expressed that more cases should be handled in this unification effort, like:

* NamedControl being capable of properly handling multiple `Control.names` issued in in one batch. This alas is somewhat hard to handle because Control.names issues bad ControlNames (and only correct Controls). Thankfully the bad ControlNames issued by `Control.names(list)` for `list.size>1` are discarded, or rather reconstructed from Controls by both the sever and by`SynthDesc as explained in this [rather long expose](https://scsynth.org/t/control-names-the-missing-manual/5007).

* Control.names.kr despite being a multi-step, state-machine process which issued names should automatically see NamedControls. I had some vague idea how this could be handled, using some caching (explained on the forum), but I did not feel like prototyping this would be useful for me personally, so I let those who desire this ["show me the code"](https://lkml.org/lkml/2000/8/25/132).

**Comment by avdrd** [comment]
For what's worth it, the bits that fix most of the uses cases I care about, meaning the 1st and 3d in the above list, are fairly easy to solve with just patches

* https://github.com/avdrd/supercollider/tree/topic/NamedControl-preload-func-args
* https://github.com/avdrd/supercollider/tree/topic/use-NamedControl-instead-of-Control.names

**Comment by telephon** [comment]
From reading the code, the solutions above look good.

**Comment by jamshark70** [comment]
> From reading the code, the solutions above look good.

I agree that it would be better to have a central repository, during SynthDef building, of all control inputs.

I think there's still a question of whether this repository should be located in NamedControl, or in SynthDef.

I tend to think SynthDef would be a better place for this.

- Option A: `arg` controls and `Control.names` controls create entries in the SynthDef, and all of them could check the entries in the SynthDef to reuse names that are declared or referenced multiple times.

- Option B: `arg` controls and `Control.names` controls create entries in the SynthDef, which are then copied over to NamedControl -- then further `arg`s (e.g. SynthDef:wrap) would need to look at NamedControl for dups.

I realize that design decisions are, to some extent, a matter of opinion, but to me, A seems cleaner.

**Comment by dyfer** [comment]
Hello, does #5675 fix this?

**Comment by telephon** [comment]
No, that fix is just a fix foe some especially painful cases in NodeProxy. 
@avdrd has made a suggestion of how to fix this here:  https://github.com/avdrd/supercollider/commit/82862d2c9634921d1b79d3ef3190abf3334ab5f4

(I just gave one suggestion about how to change the dependencies).
